msprime/ancestry.py

```diff
@@ -721,113 +762,375 @@ def simulate(
         return iterator
 
 
-# This is the initial prototype interface for sim_ancestry. A bunch of
-# parameters have been left out in the interest of simplicity so that
-# we can get an initial version in as soon as possible.
-def sim_ancestry(
-    sample_size=None,
+def _parse_rate_map(rate_param, sequence_length, name):
+    """
+    Parse the specified input rate parameter value into a rate map.
+    """
+    # Note: in the future we might have another clause here where we
+    # allow for a different  map per population. This could be
+    # accepted as either a list of N rate maps, or a dictionary mapping
+    # population names to maps.
+    # See https://github.com/tskit-dev/msprime/issues/1095
+
+    msg_head = f"Error in parsing rate map for {name}: "
+    if isinstance(rate_param, intervals.RateMap):
+        rate_map = rate_param
+        if rate_map.sequence_length != sequence_length:
+            raise ValueError(msg_head + "sequence_length must match")
+    else:
+        rate_param = 0 if rate_param is None else float(rate_param)
+        rate_map = intervals.RateMap.uniform(sequence_length, rate_param)
+    return rate_map
+
+
+def _insert_integer_samples(n, ploidy, tables):
+    """
+    Insert n individuals with the specified ploidy into the tables.
+    """
+    # Doing this with numpy is slighty obscure, but it's *much* faster
+    # than simple loops. For 10^7 samples with ploidy=2 we go from
+    # several minutes to a few seconds.
+    ind_flags = np.zeros(n, dtype=np.uint32)
+    tables.individuals.set_columns(flags=ind_flags)
+    N = n * ploidy
+    node_individual = np.zeros(N, dtype=np.int32)
+    a = np.arange(n, dtype=node_individual.dtype)
+    for j in range(ploidy):
+        node_individual[j::ploidy] = a
```

Comment:
shorter (and arguably easier to read?) is
```
node_individual = np.repeat(range(n), ploidy)
```

Reply:
Aha, yes, much better, thanks.